### PR TITLE
Reduced MonALISA Synchroniser error logs

### PR DESCRIPTION
**Pull request by xsalonx**
#### I DON'T have JIRA ticket
- [x] explain what this PR does
- [x] if it is a new feature, explain how you plan to use it
- [ ] tests are added
- [ ] documentation was updated or added

Notable changes for users:
- NA

Notable changes for developers:
- Added merging similar errors for MonALISA Synchroniser to prevent logs overflow

Changes made to the database:
- NA

### CODE REVIEW

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js [hunk 1]
```diff
@@ -71,13 +71,22 @@ class MonAlisaSynchronizer {
...
+                    .catch((error) => {
+                        if (!errors[error.message]) {
+                            errors[error.message] = [];
+                        }
+                        errors[error.message].push(monalisaDataPass.name);
+                    });
+            }
+            if (Object.entries(errors).length !== 0) {
+                this.logger.errorMessage(JSON.stringify(errors), {});
             }
```
Comment (martinboulais) [inline lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:86]: I think this is a strange way to proceed. If you have an error message very long, like an HTML error page, the error message you will display will be unreadable.
I would rather do it the following way:
Loop over the error messages, and put one log like: `Error occurred for data passes ${dataPasses.join(',')}:\n${message}`
Doing so, I would rather use a map than an object...

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js [hunk 2]
```diff
@@ -71,13 +71,22 @@ class MonAlisaSynchronizer {
                 }
             };
 
+            const errors = {};
             for (const monalisaDataPass of dataPasses) {
                 await dataSource.transaction(() => transactionablePipeline(monalisaDataPass))
```
Comment (martinboulais) [inline lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:76]: Could you rename this `transactionablePipeline` by something that inform the user about what it does?
Please think thoughtfully the name you'll use.

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js [hunk 3]
```diff
@@ -71,13 +71,22 @@ class MonAlisaSynchronizer {
...
-                    .catch((error) => this.logger.errorMessage(error.message, {}));
+                    .catch((error) => {
```
Comment (martinboulais) [inline lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:77]: You are mixing await with `.catch`, it's very strange. Please use only await.

**Review by martinboulais — COMMENTED** [review]
Please try to split the big try in smaller tries to log specific error messages

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js [hunk 4]
```diff
@@ -71,13 +71,22 @@ class MonAlisaSynchronizer {
...
+                    });
+            }
+            if (Object.entries(errors).length !== 0) {
+                this.logger.errorMessage(JSON.stringify(errors), {});
             }
             this.logger.debug('Synchronization of Data Passes completed');
         } catch (error) {
-            this.logger.errorMessage(error.message, {});
+            this.logger.errorMessage(`Fatal error preventing synchronization of DataPasses: ${error.message}`, {});
             this.logger.debug('Synchronization of Data Passes failed');
```
Comment (martinboulais) [inline lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:90]: Why is there this debug if you already logged an error message?

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js (hunk 4 above)
Comment (xsalonx) [reply lib/server/monalisa-synchronization/MonAlisaSynchronizer.js]: removed

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js (hunk 2 above)
Comment (xsalonx) [reply lib/server/monalisa-synchronization/MonAlisaSynchronizer.js]: applied

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js (hunk 3 above)
Comment (xsalonx) [reply lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:60]: applied

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js (hunk 1 above)
Comment (xsalonx) [reply lib/server/monalisa-synchronization/MonAlisaSynchronizer.js]: refactored, used your formatting. and the longest expected error there is `Unexpected token '<', "<!doctype "... is not valid JSON`

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js [hunk 5]
```diff
@@ -54,31 +54,49 @@ class MonAlisaSynchronizer {
...
-                    await dataPass.addRuns(associatedRuns, { ignoreDuplicates: true });
+            const errors = {};
+            for (const monalisaDataPass of dataPasses) {
+                try {
+                    await dataSource.transaction(() => this._updateOneDataPass(monalisaDataPass));
+                } catch (error) {
+                    if (!errors[error.message]) {
+                        errors[error.message] = [];
+                    }
+                    errors[error.message].push(monalisaDataPass.name);
```
Comment (martinboulais) [inline lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:65]:
```suggestion
                    const dataPassesNames = errors.get(error.message) ?? [];
                    if (dataPassesNames.length === 0) {
                        errors.set(error.message, dataPassesNames)
                    }
                    dataPassesNames.push(monalisaDataPass.name);
```

Comment (martinboulais) [reply lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:65]: Using a map you can do this. I would also advise to rename the variable, errors is wrong, it is `dataPassesNamesMappedByErrorMessage` (or something else, but it's not errors that you are storing)

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js [hunk 6]
```diff
@@ -54,31 +54,49 @@ class MonAlisaSynchronizer {
...
                 }
-            };
+            }
+            if (Object.entries(errors).length !== 0) {
+                const stackedError = Object.entries(errors)
+                    .map(([errorMessage, dataPassesList]) => `Error for data passes: ${dataPassesList.join(',')}\n   ${errorMessage}`)
```
Comment (martinboulais) [inline lib/server/monalisa-synchronization/MonAlisaSynchronizer.js:70]: dataPassesList => dataPassesNames once again, naming is very important

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js (hunk 5 above)
Comment (xsalonx) [reply lib/server/monalisa-synchronization/MonAlisaSynchronizer.js]: renamed

File: lib/server/monalisa-synchronization/MonAlisaSynchronizer.js (hunk 6 above)
Comment (xsalonx) [reply lib/server/monalisa-synchronization/MonAlisaSynchronizer.js]: removed

**Review by martinboulais — APPROVED** [review]

**Pull request merged**